src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,172 @@
     public T set(T value, int index) {
-        return null;
+        Node nodeToSet = getNodeByIndex(index);
+        T oldValue = nodeToSet.element;
+        nodeToSet.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node nodeToRemove = getNodeByIndex(index);
+        return unlinkNode(nodeToRemove);
     }
 
     @Override
     public boolean remove(T object) {
+        Node currentNode = head;
+        while (currentNode != null) {
+            if (object == currentNode.element || object != null
+                    && object.equals(currentNode.element)) {
+                unlinkNode(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private class Node {
+        private T element;
+        private Node prev;
+        private Node next;
+
+        private Node(T element, Node prev, Node next) {
+            this.element = element;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " out of bounds, size " + size);
+        }
+    }
+
+    private void addFirst(T elem) {
+        Node newNode = new Node(elem, null, head);
+        Node oldNode = head;
+        head = newNode;
+        if (oldNode == null) {
+            tail = newNode;
+        } else {
+            oldNode.prev = newNode;
+        }
+        size++;
+    }
+
+    private void addLast(T elem) {
+        Node newNode = new Node(elem, tail, null);
+        Node oldNode = tail;
+        tail = newNode;
+        if (oldNode == null) {
+            head = newNode;
+        } else {
+            oldNode.next = newNode;
+        }
+        size++;
+    }
+
+    private T unlinkNode(Node nodeToUnlink) {
+        final T element = nodeToUnlink.element;
+        Node prev = nodeToUnlink.prev;
+        Node next = nodeToUnlink.next;
+
+        if (prev == null) {
+            head = next;
+        } else {
+            prev.next = next;
+        }
+
+        if (next == null) {
+            tail = prev;
+        } else {
+            next.prev = prev;
+        }
+        size--;
+        return element;
+    }
+
+    private void insertBefore(Node next, T elem) {
+        Node prev = next.prev;
+        Node newNode = new Node(elem, prev, next);
+        next.prev = newNode;
+        if (prev == null) {
+            head = newNode;
+        } else {
+            prev.next = newNode;
+        }
+        size++;
+    }
+
+    private Node getNodeByIndex(int index) {
+        checkIndex(index);
+        if (size >> 1 < index) {
```

Comment:
Cool! but for better readability use arithmetic division
```suggestion
        if (size/2 < index) {
```